ref-exchange/src/stable_swap/mod.rs

```diff
@@ -0,0 +1,1092 @@
+    pub stop_amp_time: Timestamp,
+}
+
+impl StableSwapPool {
+    pub fn new(
+        id: u32,
+        token_account_ids: Vec<ValidAccountId>,
+        token_decimals: Vec<u8>,
+        amp_factor: u128,
+        total_fee: u32,
+    ) -> Self {
+        for decimal in token_decimals.clone().into_iter() {
+            assert!(decimal <= MAX_DECIMAL, "{}", ERR60_DECIMAL_ILLEGAL);
+            assert!(decimal >= MIN_DECIMAL, "{}", ERR60_DECIMAL_ILLEGAL);
+        }
+        assert!(
+            amp_factor >= MIN_AMP && amp_factor <= MAX_AMP,
+            "{}",
+            ERR61_AMP_ILLEGAL
+        );
+        assert!(total_fee < FEE_DIVISOR, "{}", ERR62_FEE_ILLEGAL);
+        Self {
+            token_account_ids: token_account_ids.iter().map(|a| a.clone().into()).collect(),
+            token_decimals,
+            amounts: vec![0u128; token_account_ids.len()],
+            volumes: vec![SwapVolume::default(); token_account_ids.len()],
+            total_fee,
+            shares: LookupMap::new(StorageKey::Shares { pool_id: id }),
+            shares_total_supply: 0,
+            init_amp_factor: amp_factor,
+            target_amp_factor: amp_factor,
+            init_amp_time: 0,
+            stop_amp_time: 0,
+        }
+    }
+
+    pub fn get_amp(&self) -> u32 {
+        if let Some(amp) = self.get_invariant().compute_amp_factor() {
+            amp as u32
+        } else {
+            0
+        }
+    }
+
+    fn get_invariant(&self) -> StableSwap {
+        StableSwap::new(
+            self.init_amp_factor,
+            self.target_amp_factor,
+            env::block_timestamp(),
+            self.init_amp_time,
+            self.stop_amp_time,
+        )
+    }
+
+    /// Returns token index for given token account_id.
+    fn token_index(&self, token_id: &AccountId) -> usize {
+        self.token_account_ids
+            .iter()
+            .position(|id| id == token_id)
+            .expect(ERR63_MISSING_TOKEN)
+    }
+
+    /// Returns given pool's total fee.
+    pub fn get_fee(&self) -> u32 {
+        self.total_fee
+    }
+
+    /// Returns volumes of the given pool.
+    pub fn get_volumes(&self) -> Vec<SwapVolume> {
+        self.volumes.clone()
+    }
+
+    /// Get per lp token price, with 1e8 precision
+    pub fn get_share_price(&self) -> u128 {
+        let mut c_current_amounts = self.amounts.clone();
+        let mut sum_token = 0_u128;
+        for (index, value) in self.token_decimals.iter().enumerate() {
+            let factor = 10_u128
+                .checked_pow((TARGET_DECIMAL - value) as u32)
+                .unwrap();
+            c_current_amounts[index] *= factor;
+            sum_token += c_current_amounts[index];
+        }
+
+        U256::from(sum_token)
+            .checked_mul(100000000.into())
+            .unwrap()
+            .checked_div(self.shares_total_supply.into())
+            .unwrap()
+            .as_u128()
+    }
+
+    pub fn predict_add_stable_liqudity(
```

Comment:
This seems to have a lot of duplicated logic with `add_liquidity`

Reply:
fixed.